### PR TITLE
Annotations.

### PR DESCRIPTION
**Pull request by untitaker**
Maybe the decorator should be applied by default.

### CODE REVIEW

**Comment by untitaker** [comment]
This is obviously not finished yet, most importantly it needs some fallback syntax for Python 2.

**Comment by DasIch** [comment]
Wow this is really nice. Having to apply the annotations decorator always seems somewhat inelegant though, although it makes sense given that it takes arguments. I'm wondering though whether we could not combine the implicit and explicit approach to applying the decorator. That is allow users to apply it, if someone wants to pass different arguments from the defaults to it and if it hasn't been applied, apply it implicitly. This way you don't have the burden of having to apply it but you don't give up control entirely.

**Comment by DasIch** [comment]
Regarding the Python 2 compatibility, how about just creating a generic decorator that sets `__annotations__` on the function?

**Comment by untitaker** [comment]
I did that now. I think more tests are needed, and i don't know how to make the style checker shut up.

**Comment by untitaker** [comment]
When i finally get this to work, i will do a `rebase -i`.

**Comment by untitaker** [comment]
I now merged #7 into this and added some more tests.

File: tests/test_annotations.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    @annotations()
+    def foo(context, bar, flag=True):
+        assert bar == 42
+        return flag
+
+    assert not foo(None, bar=42, flag='false')
+    assert foo(None, bar=42, flag='true')
+
+
+@pytest.mark.parametrize('main_code', [
```
Comment (untitaker) [inline tests/test_annotations.py]: Should i remove this code and just test the 2-3-compatible subset?

**Comment by untitaker** [comment]
Argh, this is still failing. I need to step up my game...

**Comment by untitaker** [comment]
> and if it hasn't been applied, apply it implicitly

I wonder how to detect whether a decorator already has been applied. I mean, in a way which doesn't break when other decorators are applied inbetween.

**Comment by untitaker** [comment]
- The decorator now gets applied if it hasn't been applied yet. Argvard sets the annotation for the `context` arg to a special object, to detect if the decorator has been applied multiple times. It seems `__annotations__` gets correctly set on any intermediate decorators, at least according to the unit tests.
- Added a changelog entry.

**Comment by DasIch** [comment]
Thanks you very much, this is amazing.

**Pull request merged**